stan/math/prim/mat/fun/gp_exp_quad_cov.hpp

```diff
@@ -177,50 +180,60 @@ gp_exp_quad_cov(const std::vector<T_x1>& x1, const std::vector<T_x2>& x2,
  *
  * @tparam T_x1 type of first std::vector of elements
  * @tparam T_x2 type of second std::vector of elements
- * @tparam T_sigma type of sigma
+ * @tparam T_s type of sigma
  * @tparam T_l type of length scale
  *
  * @param x1 std::vector of elements that can be used in square distance
  * @param x2 std::vector of elements that can be used in square distance
  * @param sigma standard deviation
  * @param length_scale std::vector of length scale
+ *    This function assumes the dimension if x1, x2 and l are the same.
```

Comment:
Why does it assume that this is true? What's the motivation for assuming this?

Reply:
There are columns of the design matrix. We're assuming we're not adding and dropping covariates on each iteration of the sampler. I think this is safe if we're declaring `x`  and `x_pred` as data with constant dimension.

Reply:
Since `length_scale` is an `std::vector`, that means that it's actually not a `matrix` in the Stan language. It's an array.

I don't see why `x1`, `x2`, and `l` would be assumed to be the same length. It'll definitely be user error if they aren't, but are we really saying that if they aren't, it's undefined behavior?

Reply:
Not x1, x2 the same length, the **dimension** of x1, x2

The dimension in this case is the number of columns, or we can think of it that way. In each dimension (or covariates, columns, etc), we need a seperate length scale. So computationally we'll go out of memory and hit a segfault, and it's not defined mathematically for the # of length scales to be more or less than the # of dimensions. This is also related to the buggy error handling below. So there are two cases: 1) I do the error handling, which requires I index `x` to get a row of the matrix, or 2) I can assume they're all the same length, or may be there is another solution. You dig?